sync.go

```diff
@@ -1,12 +1,207 @@
 package textsecure
 
 import (
+	"encoding/binary"
+	"fmt"
+	"io/ioutil"
+	"os"
+
 	log "github.com/Sirupsen/logrus"
+	"github.com/golang/protobuf/proto"
 	"github.com/janimo/textsecure/protobuf"
 )
 
-// handleSyncMessage handles an incoming SyncMessage
+// handleSyncMessage handles an incoming SyncMessage.
 func handleSyncMessage(src string, timestamp uint64, sm *textsecure.SyncMessage) error {
-	log.Debugf("SyncMessage is %+v\n", sm)
+	log.Debugf("SyncMessage recieved at %d", timestamp)
+	if !config.EnableMultiDeviceSync {
+		log.Debugf("Multi-Device sync is disabled. Ignoring message")
+		return nil
+	}
+
+	if sm.GetSent() != nil {
+		return handleSyncSent(sm.GetSent(), timestamp)
+	} else if sm.GetRequest() != nil {
+		return handleSyncRequest(sm.GetRequest())
+	} else if sm.GetRead() != nil {
+		return handleSyncRead(sm.GetRead())
+	} else {
+		log.Errorf("SyncMessage contains no known sync types")
+	}
+
+	return nil
+}
+
+// handleSyncSent handles sync sent messages
+func handleSyncSent(s *textsecure.SyncMessage_Sent, ts uint64) error {
+	dm := s.GetMessage()
+	dest := s.GetDestination()
+	timestamp := s.GetTimestamp()
+
+	if dm == nil {
+		return fmt.Errorf("DataMessage was nil for SyncMessage_Sent")
+	}
+
+	flags, err := handleFlags(dest, dm)
+	if err != nil {
+		return err
+	}
+
+	atts, err := handleAttachments(dm)
+	if err != nil {
+		return err
+	}
+
+	gr, err := handleGroups(dest, dm)
+	if err != nil {
+		return err
+	}
+
+	msg := &Message{
+		source:      dest,
+		message:     dm.GetBody(),
+		attachments: atts,
+		group:       gr,
+		timestamp:   timestamp,
+		flags:       flags,
+	}
+
+	if client.SyncSentHandler != nil {
+		client.SyncSentHandler(msg, ts)
+	}
+
+	return nil
+}
+
+// handleSyncRequestMessage
+func handleSyncRequest(request *textsecure.SyncMessage_Request) error {
+	if request.GetType() == textsecure.SyncMessage_Request_CONTACTS {
+		return sendContactUpdate()
+	} else if request.GetType() == textsecure.SyncMessage_Request_GROUPS {
+		return sendGroupUpdate()
+	}
+
+	return nil
+}
+
+// sendContactUpdate
+func sendContactUpdate() error {
+	log.Debugf("Sending contact SyncMessage")
+
+	lc, err := GetRegisteredContacts()
+	if err != nil {
+		return fmt.Errorf("could not get local contacts: %s", err)
+	}
+
+	tmp, err := ioutil.TempFile(config.StorageDir, "multidevice-contact-update")
+	if err != nil {
```

Comment:
Did you consider using a bytes.Buffer or a bytes.Reader instead of a tempfile in these two functions?